docs/root/overview.rst

```diff
@@ -0,0 +1,97 @@
+# Nighthawk: architecture and key concepts
+
+## High level interaction model
+
+**Process** creates one or more **Workers**. **Worker** will run **Sequencer**, which in turn queries **RateLimiter** for request-release timings. When it is time to release a request, **BenchmarkClient** will be requested to do so by **Sequencer**. **BenchMarkClient** will then ask its underlying **Pool** to create a **StreamDecoder** release the actual request. **StreamDecoder** will query the request data it needs to send from the configured **RequestSource**, and send it off. **StreamDecoder** will emit events as it progresses (pool ready, completion, etc), and timings will subsequently be recorded into **Statistic** as well as get bubbled up to **Sequencer** for tracking in-flight work and **Statistic** bookkeeping.
+
+**Sequencer** will query the configured **TerminationPredicates** to terminate when and how to terminate execution. When all **Workers** have finished, **Process** will collect the results from all **Workers** via **OutputCollector**, and run **OutputFormatter** to transform to the requested output format.
+
+## Notable upcoming changes
+
+Calling out two new concepts that may get proposed in the future, and cause some churn in the code base as we inject them.
+
+### Phases
+
+One notable addition / change that may get proposed in the near future is the introduction of **Phase**. **Phase** would represent a distinct stage of an execution, like, for example, *warmup***.** It would then be useful to have per-phase reporting of latencies as well as counters and latencies. 
+
+Concretely, a warm-up phase could be represented by a distinct duration termination predicate and a ramping rate limiter. Upon completion, the `hot` BenchmarkClient with its associated pool would then be live-transferred to the next configured phase, after which execution can continue.
+
+One other reason to have this is that it would enable remote- and/or cli- controlled ramping of certain test parameters by associating those to different phases. Termination predicates can be leveraged to immediately terminate the current phase after injecting a new one, allowing for real-time steering via gRPC and/or CLI.
+
+### Streaming parameterizations and output stats
+
+One we have phases, the gRPC service, and perhaps the CLI, would be natural candidates to follow up with to allow dynamic phase injection, as well as send back reports per phase.
+
+## Key concept descriptions
+
+### Process
+
+**Process** represents the primary entry point to a Nighthawk execution run. Only one Process is allowed to exist at the same point in time within an OS Process. **Process **is responsible for performing process-wide initialization and termination, as well as handle input configuration, deliver output, and co-ordinate Workers. **ProcessImpl **is re-used accross the CLI and the gRPC service.
```

Comment:
Would this section benefit from links to header files where the individual concepts are defined?

Reply:
Sure; however, I added a single one under "#Key concept descriptions ..." which points to the directory where they live to lower the maintenance burden on this. Are you ok with that?

Reply:
Sounds good, less links means less maintenance.